### DISCUSSION TITLE
dynamic caused the component to render twice

### DISCUSSION THREAD

**Discussion by nagq** [vercel/next.js#37631] [App Router]
### source code
```js
import dynamic from "next/dynamic";

const load = (name) => dynamic(() => import(`../components/${name}`), {});

const A = load("A");

export default function (props) {
  return (
    <>
      <A></A>
    </>
  );
}
```
### react 17 page show

```html
this is a
```

### react 18 page show

```json
// package.json
"dependencies": {
    "next": "12.1.6",
    "react": "18.1.0",
    "react-dom": "18.1.0"
}
```

```html
this is a
this is a
```

view in codesandbox: https://codesandbox.io/s/hungry-sid-8oxo7j?file=/pages/index.js

![image](https://user-images.githubusercontent.com/15318124/174234299-7d89ca82-6e15-4c59-83bc-4f88243777b2.png)

**Comment by icyJoseph** [comment]
> Tried to make sense out of this, and use a plausible fix, that I think should be the way to go anyway. Lot of trial and error on this one, apologies for its length.

For me without, initially,  it wouldn't even render at all, but your codesandbox does... and that's because I had a custom  `_app` component. 

This was a huge hint to help me understand the problem! 

If we use HTML tags instead of fragments `<></>` (nothing at all also seems to work) then the entire thing errors out... on failed hydration because of mismatched UI.

### Fix

```tsx
import { Suspense } from "react";

import type { NextPage } from "next";
import dynamic from "next/dynamic";

const load = (name: string) =>
  dynamic(() => import(`../components/${name}`), { ssr: true, suspense: true }); // both set to true

const A = load("A");

const Home: NextPage = () => {
 // By wrapping with Suspense we could even use `div's` , or have a custom _app
 // the fragments become also optional
  return (
    <>
      <Suspense fallback={null}>
        <A />
      </Suspense>
    </>
  );
};

export default Home;
```

Anyway, I think that because in React 18 suspenseful stuff is actually awaited for, there might be an issue with the server rendering the dynamic component, and then the client doing the same work again, which is how we end up rendering both.

### Cooking explanation

```tsx
import type { NextPage } from "next";
import dynamic from "next/dynamic";

const load = (name: string) =>
  dynamic(() => import(`../components/${name}`), { ssr: true });

const A = load("A");

const Home: NextPage = () => {
  return (
    <>
      <A />
    </>
  );
};

export default Home;
```

> `ssr` is true by default, I am just being verbose here.

The above renders twice, but if we inspect the page source code, that is the server generated HTML, we see only once `<A />` tag, as `<h1>this is a</h1>`. This tells me that when the client hydrates it is failing to see that the HTML for this component already exists.

In fact the React Dev Tools do not even see the server rendered `<A/>` as part of the React tree.

Wrapping the `<A/>` in Suspense does change things.

Using `ssr: false` does also help, but it is a trade off. 

> It is also worth noting that this happens, even if you do
> 
> ```js
> const A = dynamic(() => import(`../components/A`), {
>   ssr: true,
>   suspense: false,
> });
> ```

### Explanation

When at the top of the file, you use Fragments, nothing wraps the results, so they are placed as first level child to `<div id="__next"></div>`.

When the client hydrates, it fails to pick up on the already existing child, and just renders next to it. **Potentially a bug? Because it should try to hydrate that server string with `A`, right?**

By placing a Suspense boundary, we tell React that this is asynchronous rendering, and it effectively creates a layer between the root `__next` and the content. And guess what, we can see that boundary in the HTML (view page source):

```html
<div id="__next"><!--$--><h1>this is a</h1><!--/$--></div>
```

My guess is that this comment tells React as it hydrates the client that here comes a suspense boundary, which matches the source code, so it can continue working there. 

When we use the `dynamic` component raw, at the very top of the React tree we control, either as:

```tsx
const A = dynamic(() => import(`../components/A`), {
  ssr: true,
  // suspense: true,
});

const Home: NextPage = () => {
  return <A />;
};

```

or

```tsx
const A = dynamic(() => import(`../components/A`), {
  ssr: true,
  // suspense: true,
});

const Home: NextPage = () => {
  return <><A /></>;
};

```

Then, during hydration, React  "sees" a matching UI, but because dynamic, is, well, dynamic, when it resolves, the result is placed right next to the already existing UI. The question is, which one is taken as part of the React tree, as said before, the React Dev Tools, only acknowledge one. We can easily answer that question by modifying `A`:

```jsx
import { useEffect, useState } from "react";

const A = (props) => {
  const [show, setShow] = useState(false);
  useEffect(() => {
    setShow(true);
  }, []);

  return (
    <div>
      <h1>this is a</h1>

      {show ? <div>CLIENT</div> : <div>SERVER</div>}
    </div>
  );
};

export default A;
```

And the page source shows SERVER, and when rendered on the client:

```html
<div id="__next"><div><h1>this is a</h1><div>SERVER</div></div></div>
```

<img width="653" alt="Screenshot 2022-06-17 at 10 46 20" src="https://user-images.githubusercontent.com/21013447/174262626-79299647-0829-465c-94ed-56d853382110.png">

### What to make of this

I think you've found a very extreme edge case, not sure if someone should take a look at it. What should've happened is that there's only one message, `this is a CLIENT`, which we can do TODAY, if we use:

```tsx
const A = dynamic(() => import(`../components/A`), {
  ssr: true,
  suspense: true,
});

const Home: NextPage = () => {
  return (
    <Suspense fallback={null}>
      <A />
    </Suspense>
  );
};
``` 

> Again, `ssr` is set to true by default, just being verbose here

I am not sure however, if this should error out, or if React should be able to pick up on the already present HTML and hydrate it, rather than rendering next to it. When you write it like that, it does sound like something is broken.

**Reply by LaCocoRoco** [reply]
Thanks for the answer, it was a pleasure to read.

**Comment by hsheikhali1** [comment]
Hey @icyJoseph - thanks for this very detailed writeout. I am also noticing this, it's strange that my components are being "rendered twice" but when i look at the react tree there is only 1 component there.. Console log statements show that it's either rendering two of the same components, or it's the same component (1 component) being re-rendered.. 

I wonder if this has any performance hits?
